### ISSUE TITLE
v2.2 GeneralBuffer  is empty error

### ISSUE THREAD

**Issue by BookChan** [NVIDIA-Merlin/HugeCTR#21]
We run v2.2 in v100 with cuda10.1, and has some error:
[HCDEBUG][ERROR] Runtime error: GeneralBuffer is empty /tmp/HugeCTR/HugeCTR/include/general_buffer.hpp:136 

Our config is:
```json
{
    "solver": {
      "lr_policy": "fixed",
      "display":  100,
      "max_iter":  1000,
      "gpu":  [0],
      "input_key_type":"I64",
      "batchsize":  4096,
      "batchsize_eval":4096,
      "snapshot": 10000000,
      "snapshot_prefix": "./",
      "eval_interval": 100,
      "eval_metrics": ["AUC:0.9","AverageLoss"],
      "eval_batches": 500
    },
    
    "optimizer": {
      "type": "Adam",
      "global_update": true,
      "adam_hparam": {
        "learning_rate": 0.001,
        "alpha": 0.001,
        "beta1": 0.9,
        "beta2": 0.999,
        "epsilon": 0.00000001
      }
    },
  
    "layers": [ 
        {
        "name": "data",
        "type": "Data",
        "source": "./file_list.txt",
        "eval_source": "./file_list_test.txt",
        "check": "Sum",
        "label": {
          "top": "label",
          "label_dim": 1
        },
        "dense": {
          "top": "dense",
          "dense_dim": 0
        },
        "sparse": [
          {
            "top": "data1",
            "type": "DistributedSlot",
            "max_feature_num_per_sample": 100,
            "slot_num": 75
          }        
        ]
      },
  
      {
        "name": "sparse_embedding1",
        "type": "DistributedSlotSparseEmbeddingHash",
        "bottom": "data1",
        "top": "sparse_embedding1",
        "sparse_embedding_hparam": {
          "max_vocabulary_size_per_gpu": 20000000,
          "load_factor": 0.75,
          "embedding_vec_size": 16,
          "combiner": 1
        }
      },
  
      {
        "name": "reshape1",
        "type": "Reshape",
        "bottom": "sparse_embedding1",
        "top": "reshape1",
        "leading_dim": 1200
      },
  
  
      {
        "name": "concat1",
        "type": "Concat",
        "bottom": ["reshape1","dense"],
        "top": "concat1"
      },
  
      {
        "name": "slice1",
        "type": "Slice",
        "bottom": "concat1",
        "ranges": [[0,1200], [0,1200]],
        "top": ["slice11", "slice12"]
      },
  
  
      {
        "name": "multicross1",
        "type": "MultiCross",
        "bottom": "slice11",
        "top": "multicross1",
        "mc_param": {
          "num_layers": 3
        }
      },
  
      {
        "name": "fc1",
        "type": "InnerProduct",
        "bottom": "slice12",
        "top": "fc1",
         "fc_param": {
          "num_output": 256
        }
      },
  
      {
        "name": "relu1",
        "type": "ReLU",
        "bottom": "fc1",
        "top": "relu1" 
      },
        
      {
        "name": "dropout1",
        "type": "Dropout",
        "rate": 0.5,
        "bottom": "relu1",
        "top": "dropout1" 
      },
  
      {
        "name": "fc2",
        "type": "InnerProduct",
        "bottom": "dropout1",
        "top": "fc2",
         "fc_param": {
          "num_output": 128
        }
      },
  
      {
        "name": "relu2",
        "type": "ReLU",
        "bottom": "fc2",
        "top": "relu2"     
      },
  
      {
        "name": "dropout2",
        "type": "Dropout",
        "rate": 0.5,
        "bottom": "relu2",
        "top": "dropout2" 
      },
         {
        "name": "fc3",
        "type": "InnerProduct",
        "bottom": "dropout2",
        "top": "fc3",
         "fc_param": {
          "num_output": 64
        }
      },
  
      {
        "name": "relu3",
        "type": "ReLU",
        "bottom": "fc3",
        "top": "relu3"     
      },
  
      {
        "name": "dropout3",
        "type": "Dropout",
        "rate": 0.5,
        "bottom": "relu3",
        "top": "dropout3" 
      },
      {
        "name": "concat2",
        "type": "Concat",
        "bottom": ["dropout3","multicross1"],
        "top": "concat2"
      },
      
      {
        "name": "fc4",
        "type": "InnerProduct",
        "bottom": "concat2",
        "top": "fc4",
         "fc_param": {
          "num_output": 1
        }
      },
      
      {
        "name": "loss",
        "type": "BinaryCrossEntropyLoss",
        "bottom": ["fc4","label"],
        "top": "loss"
      } 
    ]
  }
```

**Comment by BookChan** [comment]
We find the  reason that cause the error:
the dense dim is 0 ,  in v2.1 , it's ok  because  label  tensor and dense tensor share the same buffer and  the model config work even if densen dim is 0.  However, in v2.2 , they use different tensor , and  when dense dim is 0 , the cudaMalloc will not work.  

There  are two ways to fix this error:
1. when  tensor dim is 0 , the GeneralBuffer should throw some error that explicitly show the error.
2. Or  when the   tensor dim is 0 ,    he GeneralBuffer  ptr_  should not be nullptr.
```c++
// general_buffer.hpp
  void init(int device_id) {
    if (initialized_ != false) CK_THROW_(Error_t::IllegalCall, "Initilized general buffer");
    device_id_ = device_id;
    CudaDeviceContext context(device_id);
    if (current_offset_ > 0) {
      CK_CUDA_THROW_(cudaMalloc((void**)&ptr_, current_offset_ * sizeof(T)));
      CK_CUDA_THROW_(cudaMemset(ptr_, 0, current_offset_ * sizeof(T)));
    }
    initialized_ = true;
  }
```

**Comment by zehuanw** [comment]
Thank you for your feedback, we have encountered this problem before but after the fix below. https://github.com/NVIDIA/HugeCTR/blob/master/HugeCTR/src/data_collector.cu#L64
It never comes up. We will try to reproduce the error.

**Comment by BookChan** [comment]
> Thank you for your feedback, we have encountered this problem before but after the fix below. https://github.com/NVIDIA/HugeCTR/blob/master/HugeCTR/src/data_collector.cu#L64
> It never comes up. We will try to reproduce the error.

The error can be reproduced  when  the dense dim is zero and  the  coming layer use the dense data( it's ok in v2.1):
```json
 {
        "name": "concat1",
        "type": "Concat",
        "bottom": ["reshape1","dense"],
        "top": "concat1"
      }
```

**Comment by zehuanw** [comment]
Thank you, we will fix this error in next patch by:
```
+++ b/HugeCTR/src/parser.cpp
@@ -99,6 +99,9 @@ InputOutputInfo get_input_tensor_and_output_name(
     if (!find_item_in_map(tensor, bstr, tensor_list)) {
       CK_THROW_(Error_t::WrongInput, "No such bottom: " + bstr);
     }
+    if(tensor->get_num_elements() <= 0){
+      CK_THROW_(Error_t::WrongInput, "Empty bottom: " + bstr);
+    }
     bottom_tensors.push_back(tensor);
   }
   return {bottom_tensors, top_strs};

```

**Comment by bashimao** [comment]
@zehuanw I think we can close this.
